keps/sig-node/20190226-migrate-in-tree-dockershim-to-out-of-tree.md

```diff
@@ -0,0 +1,148 @@
+
+> You can check [the discussion in sig-node mailing list](https://groups.google.com/forum/#!msg/kubernetes-sig-node/0qVzfugYhro/l6Au216XAgAJ) for more details. 
+
+Based on all the discussion, we agree that we should not rush to immediate decision. At the same time, it's the right time to start designing and documenting dockershim deprecation criteria and plan, which will be the main content of rest of this KEP.
+
+### Goals
+
+- A concrete dockershim deprecation criteria.
+- A brief plan to deprecate dockershim spanning multiple releases.
+- Issues need to be fixed in kubelet in order to separate dockershim from it.
+
+### Non-Goals
+
+- Deprecation of dockershim immediately without consideration for users and WIP efforts depending on it.
+- Refactoring or re-design of dockershim itself due to deprecation.
+
+## Proposal
+
+### Dockershim deprecation criteria
+
+- CRI itself is beta.
+- kubelet has no dependency on dockershim/docker in its whole lifecycle. 
+- All node related features are CRI generic and have no "back door" dependency on dockershim/docker.
+- Reasonable benchmark result of performance degradation after moving dockershim to out-of-tree.
+- A out-of-tree CRI for docker is implemented and well maintained, and become to beta.
+- E2E test framework has been updated with fully support of out-of-tree CRI container runtime.
+
+### Dockershim deprecation plan
+
+Step 1: Stabilize in-tree dockershim and decouple dockershim from kubelet (but still in-tree).
+
+Target releases: 1.15, 1,16, 1.17
+
+Actions:
+
+- Mark in-tree dockershim as "maintenance mode":
+  - CRI generic changes/features can continue on dockershim.
+  - WIP efforts on dockershim can continue and go to complete.
+  - dockershim/docker specific changes/features should be rejected.
+- Revisit and fix dependency issues of dockershim/docker in kubelet (e.g. #54405). Currently, kubelet still has:
+  - vendored dockershim 
+  - flags that are used to configure dockershim.
+  - support to get container logs when docker uses journald as the driver.
+  - logic of moving docker processes to a given cgroup
+  - TBD anything else?
+- Package in-tree dockershim separately from kubelet and provide a "option" to enable/disable it.
+- Ensure e2e/Node e2e test framework is CRI generic and test cases are independent of container runtime.
+
+Step 2: Work out a out-of-tree CRI for docker
+
+Target releases: 1.18
+
+Actions:
+
+- Design & implement a out-of-tree CRI for docker, it can be "copied" from dockershim as beginning.
+- Benchmark performance impact of out-of-tree CRI for docker.
+- Re-direct dockershim related features/changes to this out-of-tree CRI for docker.
+
+
+Step 3: Completely deprecate in-tree dockershim from kubelet.
+
+Target releases: TBD, we probably need to continue keeping in-tree dockershim for 3 releases as grace period.
+
+Actions:
+
+- Refactoring e2e/Node e2e test framework to include CRI for docker installation (or use other CRI container runtime).
```

Comment:
this would be extremely helpful for existing CRI impls to test/verify compatibility

Reply:
Cluster/node e2e doesn't care what CRI implementation you use under the hood. I think maybe I've missed the point of this statement.

Reply:
There're two points in my mind: 
1. Ensure cluster/node e2e are 100% CRI focused. For instance: https://github.com/kubernetes/kubernetes/issues/55414
2. Ensure test-infra install CRI-docker or cri-containerd binary in e2e machines. Currently, they install Docker binary only.